docs/dev/design/better-doc-urls-handling.rst

```diff
@@ -0,0 +1,417 @@
+Better handling of docs URLs
+============================
+
+``proxito`` is the component of our code base in charge of serving documentation
+to users and handling any other URLs from the user documentation domain.
+
+The current implementation has some problems that are discussed in this document,
+and an alternative implementation is proposed to solve those problems.
+
+Current implementation
+----------------------
+
+The current implementation is based on Django URLs
+trying to match a pattern that looks like a single project, a versioned project,
+or a subproject, this means that a couple or URLs are *reserved*,
+and won't resolve to the correct file if it exists
+(https://github.com/readthedocs/readthedocs.org/issues/8399, https://github.com/readthedocs/readthedocs.org/issues/2292),
+this usually happens with single version projects.
+
+And to support custom URLs we are hacking into Django's urlconf
+to override it at runtime,
+this doesn't allow us to implement custom URLs for subprojects easily
+(https://github.com/readthedocs/readthedocs.org/pull/8327).
+
+Alternative implementation
+--------------------------
+
+Instead of trying to map a URL to a view,
+we first analyze the canonical project (given from the subdomain),
+and based on that we map each part of the URL (parts are the result of splitting the URL on ``/``)
+to the *current* project.
+
+This will allow us to re-use this code in our resolver
+without the need to override the Django's urlconf at runtime,
+or guessing a project only by the structure of its URL.
+
+Terminology:
+
+Canonical project
+  The project from where the documentation
+  is served (the parent project of a subproject or translation).
+Current project
+  The project that owns the current file being served
+  (a subproject, a translation, etc).
+URL part
+  The result of splitting the URL on ``/``.
+Requested file
+  The final path to the file that we need to serve from the current project.
```

Comment:
This is wonderful. Really appreciate the defined terms -- we should add these to a `dev` glossary, and standardize our use of them.